src/nvim/eval.c

```diff
@@ -19754,16 +19815,51 @@ char_u *do_string_sub(char_u *str, char_u *pat, char_u *sub, char_u *flags)
   return ret;
 }
 
+static inline JobOptions common_job_options(char **argv, char *autocmd_file)
+{
+  TerminalJobData *data = xcalloc(1, sizeof(TerminalJobData));
+  if (autocmd_file) {
+    data->autocmd_file = xstrdup(autocmd_file);
+  }
+  JobOptions opts = JOB_OPTIONS_INIT;
+  opts.argv = argv;
+  opts.data = data;
+  opts.stdout_cb = on_job_stdout;
+  opts.stderr_cb = on_job_stderr;
+  opts.exit_cb = on_job_exit;
+  return opts;
+}
+
+static inline Job *common_job_start(JobOptions opts, typval_T *rettv)
+{
+  Job *job = job_start(opts, &rettv->vval.v_number);
+  TerminalJobData *data = opts.data;
+  data->refcount++;
```

Comment:
* We need to increase `data->refcount` before `job_start()`, because the clean up code for a failed `process_spawn()` will also call `term_job_data_decref()`. 
* We use the wrong pointers for the handles in the `uv_close()` calls when `process_spawn()` fails. 

```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 74a1915..0246cb7 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -19832,18 +19832,19 @@ static inline JobOptions common_job_options(char **argv, char *autocmd_file)
 
 static inline Job *common_job_start(JobOptions opts, typval_T *rettv)
 {
-  Job *job = job_start(opts, &rettv->vval.v_number);
   TerminalJobData *data = opts.data;
   data->refcount++;
+  Job *job = job_start(opts, &rettv->vval.v_number);
 
   if (rettv->vval.v_number <= 0) {
     if (rettv->vval.v_number == 0) {
       EMSG(_(e_jobtblfull));
       free(data->autocmd_file);
+      free(data);
     } else {
       EMSG(_(e_jobexe));
-      free(opts.data);
     }
+    free(opts.term_name);
     return NULL;
   }
   data->job = job;
diff --git a/src/nvim/os/job.c b/src/nvim/os/job.c
index 40459d7..12006b7 100644
--- a/src/nvim/os/job.c
+++ b/src/nvim/os/job.c
@@ -142,13 +142,13 @@ Job *job_start(JobOptions opts, int *status)
   // Spawn the job
   if (!process_spawn(job)) {
     if (opts.writable) {
-      uv_close((uv_handle_t *)&job->proc_stdin, close_cb);
+      uv_close((uv_handle_t *)job->proc_stdin, close_cb);
     }
     if (opts.stdout_cb) {
-      uv_close((uv_handle_t *)&job->proc_stdout, close_cb);
+      uv_close((uv_handle_t *)job->proc_stdout, close_cb);
     }
     if (opts.stderr_cb) {
-      uv_close((uv_handle_t *)&job->proc_stderr, close_cb);
+      uv_close((uv_handle_t *)job->proc_stderr, close_cb);
     }
     process_close(job);
     event_poll(0);
```